### ISSUE TITLE
Disabling SSO causes default Java EE Security authentication mechanisms stop working

### ISSUE THREAD

**Issue by FabriSr** [OpenLiberty/open-liberty#17959]
Disabling SSO (setting singleSignonEnabled attribute of webAppSecurity element to false ) causes the default Java EE Security authentication mechanisms stop working. 

**Steps to Reproduce**

You can run the finished project of the _[Securing a web application guide](https://openliberty.io/guides/security-intro.html#try-what-youll-build)_. Before run, in server.xml, set the singleSignonEnabled attribute of the webAppSecurity element to false. In the login page enter the username "bob" and password "bobpwd" and click Login. The application just refresh the login page. Turning on SSO the login page works as expected.

The same occurs in this [demo project](https://github.com/OpenLiberty/open-liberty/files/6874604/openliberty_error_demo.zip) using CustomFormAuthenticationMechanismDefinition. Run the application and enter any user and password and click in "login". 
In the stack trace, it shows:

`Error: SESN0008E: A user authenticated as anonymous has attempted to access a session owned by user.`

**Expected behavior**
User login.

**Diagnostic information:**
 - OpenLiberty Version:21.0.0.7
 - Java Version: OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)

**Comment by arkarkala** [comment]
I'll admit that this property (singleSignonEnabled) should be documented better in Liberty. Here is the info from the tWAS docs which explains this in more detail

_Web applications that use Java EE FormLogin style login pages, such as the administrative console, require single sign-on (SSO) enablement. Only disable SSO for certain advanced configurations where LTPA SSO-type cookies are not required_

So in your case since you are using formLogin/customformLogin, this needs to be enabled for ltpaToken to be created and used for form authentication to work. Only in rare situations this can be disabled - some info here -   https://stackoverflow.com/questions/53524601/disabling-ltpa-things-in-a-liberty-server-am-i-missing-something

However, in your case, since you seem to use JSR375 (Java EE Security API), have you looked into the RememberMe annotation (https://javaee.github.io/security-api/apidocs/javax/security/enterprise/authentication/mechanism/http/RememberMe.html) and the corresponding RememberMeIdentityStore (https://javaee.github.io/security-api/apidocs/javax/security/enterprise/identitystore/RememberMeIdentityStore.html) that might potentially help your use case? When the RememberMeIdentityStore is used it will create the loginToken and also validate it - so the control is all in the implementation (the token here will take precedence over the ltpaToken).

**Comment by FabriSr** [comment]
> Web applications that use Java EE FormLogin style login pages, such as the administrative console, require single sign-on (SSO) enablement. Only disable SSO for certain advanced configurations where LTPA SSO-type cookies are not required.

> So in your case since you are using formLogin/customformLogin, this needs to be enabled for ltpaToken to be created and used for form authentication to work. Only in rare situations this can be disabled - some info here - https://stackoverflow.com/questions/53524601/disabling-ltpa-things-in-a-liberty-server-am-i-missing-something

So, this behavior I reported is not a bug? To fully implement JSR 375 (which includes implementing basic authentication mechanisms and features like @AutoApplySession as required by JSR 375) does Liberty need SSO?

> When the RememberMeIdentityStore is used it will create the loginToken and also validate it - so the control is all in the implementation (the token here will take precedence over the ltpaToken).

The problem is that, in order to use the @RememberMe feature, I will have to create and maintain a custom authentication mechanism when I just need to avoid SSO (isolate authentication). The mechanisms provided in JSR 375 meet my needs (eg @CustomFormAuthenticationMechanismDefinition).


So, in my opnion, there are 3 issues related to the SSO feature:

- The singleSignonEnabled setting is poorly documented
- Liberty requires SSO to provide all the features required in JSR 375
- The settings related to the cookie domain are ineffective from a security point of view as, on the server side, the server does not monitor the generated tokens against the cookie domain. Therefore, even if cookies are generated with different domains, the token itself is valid on any domain that is running on the same server. Basically, all applications are vulnerable to the pass-the-cookie attack. In my opinion, the documentation not alerting to this is a pretty serious flaw.

I believe the best solution would be to have a feature to be able to configure the scope of a token trough authentication filters. Or maybe, "apply domains to the token itself" at server side when settings like ssoUseDomainFromURL are used.

**Comment by arkarkala** [comment]
Yes, this property is working as designed wrt to ltpa and formLogin scenario. Agree that the doc improvement is needed. 

In regards to SSO when using LTPA, as the docs say, all you need is to share the ltpa keys (and UR if diff. servers). If a ltpa token is presented that can be validated with the right keys, ltpa authentication will work. If isolation is needed when using ltpa, the apps have to deployed in different servers. This will provide better isolation as you one can also have different URs, JAAS etc as needed. Ofcourse, one can also use other auth mechs (OIDC, SAML, SPNEGO etc) or custom auth (TAI, RememberMe etc) to achieve SSO.

If separation of tokens are needed in the same server, yes, there should be a way to associate the ltpa keys for different app(s) and AuthFilter would be a good place. However, this would be a new feature request.

We can use this issue to improve the doc. Thanks.
